### PR TITLE
Update for new libplacebo tone mapping / colorspace refactor

### PR DESCRIPTION
**Pull request by haasn**
https://code.videolan.org/videolan/libplacebo/-/merge_requests/212

### CODE REVIEW

**Comment by haasn** [comment]
(Note: I didn't bother mapping the LUT size because 256 works well enough for everything I've thrown at it.)

**Comment by quietvoid** [comment]
Are there plans to add control for the tonemapping modes?  (RGB, hybrid, luma, etc.)

**Comment by haasn** [comment]
Oh, yes, absolutely. I forgot about those.

**Comment by epicfilmmaker93** [comment]
I briefly tested this and SDR to HDR seems to work, however target-peak when doing both HDR to SDR and SDR to HDR conversion is not working as expected - higher value brightens the picture. Am I doing something wrong?

File: video/out/vo_gpu_next.c [hunk 1]
```diff
@@ -1396,27 +1386,36 @@ static void update_render_options(struct priv *p)
...
+        [TONE_MAPPING_AUTO]     = &pl_tone_map_auto,
+        [TONE_MAPPING_CLIP]     = &pl_tone_map_clip,
+        [TONE_MAPPING_MOBIUS]   = &pl_tone_map_mobius,
+        [TONE_MAPPING_REINHARD] = &pl_tone_map_reinhard,
+        [TONE_MAPPING_HABLE]    = &pl_tone_map_hable,
+        [TONE_MAPPING_GAMMA]    = &pl_tone_map_gamma,
+        [TONE_MAPPING_LINEAR]   = &pl_tone_map_linear,
+        [TONE_MAPPING_SPLINE]   = &pl_tone_map_spline,
+        [TONE_MAPPING_BT_2390]  = &pl_tone_map_bt2390,
+        [TONE_MAPPING_BT_2246A] = &pl_tone_map_bt2446a,
```
Comment (quietvoid) [inline video/out/vo_gpu_next.c:1399]: Pretty much all the references in mpv are typo'd as 2246 instead of 2446.

Comment (haasn) [reply video/out/vo_gpu_next.c]: Fixed, thanks for noticing

**Comment by haasn** [comment]
> Are there plans to add control for the tonemapping modes? (RGB, hybrid, luma, etc.)

Updated with these options included. I decided to also scrap `tone-mapping-desaturate` etc. in favor of also making `vo_gpu` use `--tone-mapping-mode`, to simplify things. This rather elegantly avoids any significant deviation in the way these options work between `vo_gpu` and `vo_gpu_next`, providing a nicer experience for everybody.

**Comment by haasn** [comment]
> Edit: seems to be the case only when inverse tone mapping option is used. Does it mean that target-peak is the target peak of the video in this case? Shouldn't it be renamed then? And when inverse tone mapping is disabled, there's still no change in brightness for HDR content when modifying the target-peak

`target-peak` is the peak brightness of the output display. Doesn't matter whether it's HDR or SDR, in all cases the intent is for it to be set it to the (measured) brightness of the display device.

**Comment by NSQY** [comment]
haasn, first of all thanks for revisiting this. I am finding the new features and approach very favorable.

I do have a question however, is this code making use of allow_delayed_peak_detect in libplacebo or something similar? Occasionally, I am noticing that the void is invading my media. It seems most common with sudden flashes.

With the latest mpv and libplacebo master as of writing ``` mpv --no-config --vo=gpu-next --pause --mute test.mkv --target-peak=300 --tone-mapping=bt.2446a --tone-mapping-mode=rgb --gamut-mapping-mode=desaturate --hdr-compute-peak=yes ```
Sample here: https://0x0.st/ozNO.mkv

void_harry

**Comment by epicfilmmaker93** [comment]
@haasn then it's my mistake, sorry. What about target-peak for HDR to SDR conversion? Still does nothing for me

**Comment by haasn** [comment]
> Occasionally, I am noticing that the void is invading my media. It seems most common with sudden flashes.

Fixed in haasn/libplacebo@478623de, was caused by `--tone-mapping-mode=desaturate`.

> I do have a question however, is this code making use of allow_delayed_peak_detect in libplacebo or something similar?

That's controlled by the existing option `--allow-delayed-peak-detect`.

**Comment by NSQY** [comment]
```--target-peak``` works with HDR -> SDR mapping for me also.

**Comment by epicfilmmaker93** [comment]
Ok, the problem was icc-profile-auto. If I use an ICC profile, do I even need to use target-peak? Because if so, setting it to a high value clips the highlights, and setting it to 100 makes everything darker.

Also, tone-mapping-mode is not documented in man

**Comment by quietvoid** [comment]
@epicfilmmaker93 Normally I think `target-peak` should be set to the display's peak brightness.
However the libplacebo refactor does have a regression when using an ICC profile currently.

For example, this is what I've noticed: https://slow.pics/c/KBqUlhel

**Comment by haasn** [comment]
> Edit: I definitely need target-peak, but how do I know which value should I choose so that it works well with my ICC profile?

I'm actually planning on changing the ICC handling code in libplacebo to read the true colorimetry values directly from the profile tagging (`wtpt`, `bkpt`, `{r,g,b}XYZ` and `lumi`) so we can calculate exact contrast and gamut volume information. That way we can do appropriate gamut mapping as well.

**Comment by epicfilmmaker93** [comment]
Ok, thanks guys, guess I'll stick to SDR with ICC profiles for now

**Comment by haasn** [comment]
I've decided that the ICC profile changes will take more time, and don't affect mpv anyway. So I'm comfortable merging this as-is for the time being, to get the new API in.

Any objections?

**Comment by Hrxn** [comment]
SHIP IT

**Comment by epicfilmmaker93** [comment]
I think it's better to merge this as is, maybe just print a warning to users with icc-profile-auto

**Comment by haasn** [comment]
> I think it's better to merge this as is, maybe just print a warning to users with icc-profile-auto

Isn't it just specifically the combination of icc profile + target-peak? (Judging by @quietvoid's report)

**Comment by epicfilmmaker93** [comment]
Well, then print a warning to users with both options enabled

**Comment by quietvoid** [comment]
> Isn't it just specifically the combination of icc profile + target-peak?
Pretty much.

Using an ICC profile with target-peak=203 should be the same as without a profile.
So I'd recommend using that, it might already be the default.

**Comment by pillepalle1234** [comment]
I still think bt. 2446a looks weird. While preserving more Details in Highlights it dims the picture and adds a grey fog.
target-prim=dci-p3
target-trc=gamma2.2
hdr-compute-peak=yes
tone-mapping=bt.2446a or bt.2390
tone-mapping-mode=auto

Some comparison shots:
target-peak=auto

https://ibb.co/DVPvj1D
https://ibb.co/bmfZ3yg
https://ibb.co/wd6XdH7
https://ibb.co/BT1xm5f
https://ibb.co/RcwVPtj
https://ibb.co/qx4PF2x
https://ibb.co/94ZRMHx
https://ibb.co/tcGSCrc

Target peak is somewhat broken with bt.2446a aswell i think...
target-peak= from auto till 10 (!)

https://ibb.co/94ZRMHx
https://ibb.co/7b1JVy0
https://ibb.co/qW67CjZ
https://ibb.co/PTCw8H8
https://ibb.co/tcGSCrc

**Pull request merged**